lib/retryable.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+# Wrapper for a retryable MongoDB operation on clusters. Can either create a
+# simple retryable operation or a retryable operation wrapped in a transaction.
+class Retryable
+  MAX_RETRIES=5
+  # Not constantized by mongo gem
+  MONGO_DUPLICATE_KEY_ERROR=11_000
+
+  def self.ensure_transaction
+    if (s = Mongoid::Threaded.get_session)
+      raise "In a session but not in a transaction??" unless s.in_transaction?
+
+      yield
+    else
+      Cluster.with_session do |session|
+        session.with_transaction { new.run { yield } }
```

Comment:
Something funny is going on here -- `with_transaction` ends up wrapping it in two `Retryable` blocks. I think the intention was that `ensure_transaction` would wrap the block in a transaction if needed but not do retryable part. I will change `ensure_transaction`